pandas/_libs/tslibs/parsing.pyx

```diff
@@ -188,19 +207,27 @@ cdef parse_datetime_string_with_reso(date_string, freq=None, dayfirst=False,
 
 
 cpdef bint _does_string_look_like_datetime(object date_string):
-    if date_string.startswith('0'):
-        # Strings starting with 0 are more consistent with a
-        # date-like string than a number
-        return True
-
-    try:
-        if float(date_string) < 1000:
-            return False
-    except ValueError:
-        pass
+    cdef:
+        char *buf = NULL
+        char *endptr = NULL
+        Py_ssize_t length = -1
+        double converted_date
+        char first
 
-    if date_string in _not_datelike_strings:
+    if not get_string_data(date_string, &buf, &length):
```

Comment:
having a cython routine for this just adds extra code with very little / no perf gain.

Reply:
Are you saying that having this two-function juggle to properly extract a `char *`  from Python object is _easier_?
Here's what current pandas use:
https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/np_datetime.pyx#L174

```cython
cdef inline int _string_to_dts(object val, npy_datetimestruct* dts,
                               int* out_local, int* out_tzoffset) except? -1:
    cdef:
        int result
        char *tmp

    if isinstance(val, unicode):
        val = PyUnicode_AsASCIIString(val)

    tmp = val
    result = _cstring_to_dts(tmp, len(val), dts, out_local, out_tzoffset)

    if result == -1:
        raise ValueError('Unable to parse %s' % str(val))
    return result


cdef inline int _cstring_to_dts(char *val, int length,
                                npy_datetimestruct* dts,
                                int* out_local, int* out_tzoffset) except? -1:
    # Note: without this "extra layer" between _string_to_dts
    # and parse_iso_8601_datetime, calling _string_to_dts raises
    # `SystemError: <class 'str'> returned a result with an error set`
    # in Python3
    cdef:
        int result

    result = parse_iso_8601_datetime(val, length,
                                     dts, out_local, out_tzoffset)
    return result
```

I'm not even mentioning that this approach allocates extra memory for translating `PyUnicode` to `PyBytes` even though in most cases it's not needed as unicode is mostly already in `utf-8` (and also Cython `char* foo = python_str_object` allocates extra memory in the meantime), so for a typical case when utf-8 unicode object comes in it's _two_ memory allocations that _are not needed_ whatsoever.

Here's (untested) version that uses our proposed routine for calling same function:

```cython
from cpython cimport PyUnicode_Check, PyBytes_Check, PyBytes_AsStringAndSize
cdef extern from *:
    char* PyUnicode_AsUTF8AndSize(object unicode, Py_ssize_t* length)

cdef inline bint get_string_data(object s, char **buf, Py_ssize_t *length):
    if PyUnicode_Check(s):
        buf[0] = PyUnicode_AsUTF8AndSize(s, length)
        return buf[0] != NULL
    if PyBytes_Check(s):
        return PyBytes_AsStringAndSize(s, buf, length) == 0
    return False

cdef inline int _string_to_dts(object val, npy_datetimestruct* dts,
                               int* out_local, int* out_tzoffset) except? -1:
    cdef:
        Py_ssize_t length
        char *tmp

    if not get_string_data(val, &tmp, &len):
        raise ValueError('Unable to parse %s' % str(val))
    return parse_iso_8601_datetime(tmp, length,
                                     dts, out_local, out_tzoffset)
```

Note that this `get_string_data` should be re-used everywhere where getting raw `char*` from an object is needed, thus _lessening_ the complexity of the codebase when used everywhere.

Reply:
> Note that this get_string_data should be re-used everywhere where getting raw char* from an object is needed, thus lessening the complexity of the codebase when used everywhere.

that would be better. The basic problem I have here is you are adding a ton of code which is all new, barely using anything that is existing. This leads to a huge in crease in code complexiity for  modest increase in parsing speed.

Reply:
You point about unnecessary copies being made is worth looking into separately (i.e. when I’m not typing with thumbs).   Can it be addressed in its own issue/PR?

Reply:
> Can it be addressed in its own issue/PR?

We can make a PR that introduces this util function and switch some already existing functions to it.
I doubt it alone would make significant benchmark changes, though - most benchmarks are not memory-intensive enough, and to make them intensive means they would need to work with **huge** datasets.

That said, I'm all for making this utility funciton a separate PR as we're planning to use it in at least 2 more places, but I probably won't be able to back this separate PR with performance increase.

Reply:
Extracted the function to separate PR #25895